client/session/session_storage.go

```diff
@@ -0,0 +1,116 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package session
+
+import (
+	"fmt"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/mysteriumnetwork/node/client"
+	stats_dto "github.com/mysteriumnetwork/node/client/stats/dto"
+	"github.com/mysteriumnetwork/node/core/connection"
+	"github.com/mysteriumnetwork/node/identity"
+	node_session "github.com/mysteriumnetwork/node/session"
+)
+
+const sessionStorageLogPrefix = "[session-storage] "
+
+// StatsRetriever can fetch current session stats
+type StatsRetriever interface {
+	Retrieve() stats_dto.SessionStats
+}
+
+// SessionStorer allows us to get all sessions, save and update them
+type SessionStorer interface {
+	Save(object interface{}) error
+	Update(object interface{}) error
+	GetAll(array interface{}) error
+}
+
+// SessionStorage contains functions for storing, getting session objects
+type SessionStorage struct {
+	storage        SessionStorer
+	statsRetriever StatsRetriever
+}
+
+// NewSessionStorage creates session repository with given dependencies
+func NewSessionStorage(storage SessionStorer, statsRetriever StatsRetriever) *SessionStorage {
+	return &SessionStorage{
+		storage:        storage,
+		statsRetriever: statsRetriever,
+	}
+}
+
+// Save saves a new session
+func (repo *SessionStorage) Save(se Session) error {
+	return repo.storage.Save(&se)
+}
+
+// Update updates specified fields of existing session by id
+func (repo *SessionStorage) Update(sessionID node_session.ID, updated time.Time, dataStats stats_dto.SessionStats, status SessionStatus) error {
+	// update fields by sessionID
+	se := Session{SessionID: sessionID, Updated: updated, DataStats: dataStats, Status: status}
+	return repo.storage.Update(&se)
+}
+
+// GetAll returns array of all sessions
+func (repo *SessionStorage) GetAll() ([]Session, error) {
+	var sessions []Session
+	err := repo.storage.GetAll(&sessions)
+	if err != nil {
+		return nil, err
+	}
+	return sessions, nil
+}
+
+// Subscribe subscribes the storage on the bus for relevant events
+func (repo *SessionStorage) Subscribe(bus client.EventSubscriptionKeeper) {
+	bus.Subscribe(string(connection.StateEvent), repo.consumeStateEvent)
+}
+
+// Unsubscribe unsubscribes the storage from bus
+func (repo *SessionStorage) Unsubscribe(bus client.EventSubscriptionKeeper) {
+	bus.Unsubscribe(string(connection.StateEvent), repo.consumeStateEvent)
```

Comment:
`bus.Unsubscribe` returns an error, should we check it somehow?

Reply:
These will now be checked in DI.